### DISCUSSION TITLE
Synchronous bindings?

### DISCUSSION THREAD

**Discussion by uvizhe** [fzyzcjy/flutter_rust_bridge#312] [Q&A] [answered]
Hi, thanks for this wonderful project. I'm relatively new to Rust and a very beginner in Dart/Flutter. I've a few questions regarding asynchronous nature of bindings this tool generates.

First, I quite don't understand why all the generated bindings are asynchronous and what should I do about it if I don't want or need to add async code in my Flutter UI? I've found a `SyncResult` type but it works only for `Vec<u8>` and I'm not sure why? My first thought was to encode/decode everything with msgpack (and not to deal with differences of unicode string encoding of the two languages) but this seems overcomplicated and just wrong since (I suppose) it all works out of the box with async code. Is there something I missed in the docs or synchronous bindings are just not implemented yet?

**Comment by fzyzcjy** [comment, marked as answer]
> why all the generated bindings are asynchronous and what should I do about it if I don't want or need to add async code in my Flutter UI

flutter is single threaded. if your rust code runs for longer time, e.g. even 50ms, it will block the main thread and UI will stuck.

> I've found a SyncResult type but it works only for Vec<u8> and I'm not sure why

because sync data handling is quite different and you are welcomed to make a pr to enhance it!

> synchronous bindings are just not implemented yet?

yes

**Comment by uvizhe** [comment]
> flutter is single threaded. if your rust code runs for longer time, e.g. even 50ms, it will block the main thread and UI will stuck.

Hmm, and this applies to a pure Dart code as well, if I understand it correctly?

**Comment by fzyzcjy** [comment]
Yes (mostly). Dart isolates are a bit unlike normal threads.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
